marlowe-playground-client/src/JavascriptEditor/State.purs

```diff
@@ -45,6 +48,18 @@ handleAction ::
   HalogenM State Action ChildSlots Void m Unit
 handleAction _ (HandleEditorMessage (Monaco.TextChanged text)) =
   ( do
+      -- TODO: This handler manages the logic of having a restricted range that cannot be modified. But the
+      --       current implementation uses editorSetValue to overwrite the editor contents with the last
+      --       correct value (taken from local storage). By using editorSetValue inside the TextChanged handler
+      --       the events get fired multiple times on init, which makes hasUnsavedChanges always true for a new
+      --       JS project or a project load.
+      --
+      --       Once the PR 2498 gets merged, I want to try changing the web-commons Monaco component so that the
+      --       TextChanged handler returns an IModelContentChangedEvent instead of a string. That event cointains
+      --       information of the range of the modifications, and if the action was triggered by an undo/redo
+      --       action. With that information we can reimplement this by firing an undo event if a "read only"
+      --       decoration. At this moment I'm not sure if that will solve the bubble problem but at least it will
+      --       allow us to decouple from local storage.
```

Comment:
Bubble problem?

Reply:
With bubble problem I meant that handling the action and modifying the editors value ends up firing the action again.

In this capture I logged the start and end of the TextChanged action when I reload the home page. and I also added the withLogs helper to the get more context. 
<img width="639" alt="Screen Shot 2020-12-09 at 10 19 09" src="https://user-images.githubusercontent.com/2634059/101634906-11571080-3a08-11eb-98ac-4e87151f75e4.png">

The action get's called 7 times for JS, while for the others is called 2 times. 

As I mentioned in the comment, by modifying the event in the webcommons we can get avoid using the local storage and have more information to decide when to call undo. But that might not solve the problem of tracking the unsaved status and I may have to do a full text comparison to check for that. I'd like to avoid having the fulltext comparison both for performance issues and also because that means I need to track the last saved code.

Reply:
Ah, I see, nice, yes, my solution was very inefficient. It seems there use to be a hidden functionality for locking paragraphs but they removed it in the latest versions, such a pitty.